tests/network/mod.rs

```diff
@@ -596,21 +601,43 @@ where
     D: DistAlgorithm<Output = Batch<C, NodeId>, NodeId = NodeId>,
     C: Contribution + Clone,
 {
-    /// Verifies that all nodes' outputs agree.
+    /// Verifies that all nodes' outputs agree, noting that node 0 was removed - hence could have
+    /// missed some batches - and then rejoined the network.
     #[allow(unused)] // Not used in all tests.
     pub fn verify_batches(&self) {
-        let expected = self.nodes[&NodeId(0)].outputs().to_vec();
+        let expected = self.nodes[&NodeId(1)].outputs().to_vec();
         assert!(!expected.is_empty());
         let pub_eq = |(b0, b1): (&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
-        for node in self.nodes.values() {
-            assert_eq!(expected.len(), node.outputs().len());
-            assert!(
-                expected.iter().zip(node.outputs()).all(pub_eq),
-                "Outputs of nodes 0 and {} differ: {:?} != {:?}",
-                node.instance().our_id().0,
-                expected,
-                node.outputs()
-            );
+        let pub_eq_ref = |(b0, b1): &(&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
+        for node in self.nodes.values().filter(|node| node.id != NodeId(1)) {
+            if node.id == NodeId(0) {
+                assert!(expected.len() >= node.outputs().len());
+                assert!(
+                    expected
+                        .iter()
+                        .zip(node.outputs())
+                        .take_while(pub_eq_ref)
+                        .next()
+                        .is_some(),
+                    "Outputs of nodes 0 and 1 don't start the same",
+                );
```

Comment:
That only asserts that the first batch is the same.

Reply:
I know. That's not ideal. This check should resemble `VirtualNet::vefify_batches` more.

Reply:
It turns out that the removed node still outputs a batch! Found after modifying `TestNetwork::verify_batches` to compute the same expectation as in `VirtualNet::verify_batches`. So, the test needs to be corrected again.

Reply:
I'd expect that it outputs all the blocks up to and including the one with the `Complete(…)` that removed it, and then again all blocks starting with the `InProgress(…)` that added it?

Reply:
Yes, with the random scheduler. But with the first scheduler it outputs all batches, including the one made when it wasn't a participant.

Reply:
That sounds like a bug. :thinking:

Reply:
> I'd expect that it outputs all the blocks up to and including the one with the Complete(…) that removed it, and then again all blocks starting with the InProgress(…) that added it?

All up to `Complete`, then skipping until and including `InProgress`, then all the rest? It isn't a participant in the epoch which produced the batch with `InProgress`.

Reply:
Ah, sorry, you're right, of course! The `InProgress` that adds it back is the last batch that it _doesn't_ receive. :+1:

Reply:
That assertion is still weaker than even this:
```rust
assert_eq!(expected[0], node.outputs()[0]);
```